### PR TITLE
Group file source edits by FileId

### PR DESCRIPTION
**Pull request by Veykril**

### CODE REVIEW

File: crates/assists/src/tests.rs [hunk 1]
```diff
@@ -80,10 +80,8 @@ fn check_doc_test(assist_id: &str, before: &str, after: &str) {
     let actual = {
         let source_change = assist.source_change.unwrap();
         let mut actual = before;
-        for source_file_edit in source_change.source_file_edits {
-            if source_file_edit.file_id == file_id {
-                source_file_edit.edit.apply(&mut actual)
-            }
+        if let Some(source_file_edit) = source_change.source_file_edits.edits.get(&file_id) {
```
Comment (matklad) [inline crates/assists/src/tests.rs:83]:
```suggestion
        if let Some(source_file_edit) = source_change.source_file_edits.get(file_id) {
```

I generally find time spend on nice interfaces to be spend well

File: crates/ide_db/src/source_change.rs [hunk 2]
```diff
@@ -17,27 +23,51 @@ impl SourceChange {
...
 }
 
-#[derive(Debug, Clone)]
-pub struct SourceFileEdit {
-    pub file_id: FileId,
-    pub edit: TextEdit,
+#[derive(Default, Debug, Clone)]
+pub struct SourceFileEdits {
+    pub edits: FxHashMap<FileId, TextEdit>,
+}
```
Comment (matklad) [inline crates/ide_db/src/source_change.rs:36]: Does this type pull its weight? Seems like we can just use `SourceChange` directly, no?

File: crates/ide_db/src/source_change.rs [hunk 3]
```diff
@@ -17,27 +23,51 @@ impl SourceChange {
...
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.edits.is_empty()
+    }
+
+    pub fn insert(&mut self, file_id: FileId, edit: TextEdit) {
+        match self.edits.entry(file_id) {
+            Entry::Occupied(mut entry) => {
+                entry.get_mut().union(edit).expect("overlapping edits for same file");
```
Comment (matklad) [inline crates/ide_db/src/source_change.rs:54]:
```suggestion
                assert_never!(entry.get_mut().union(edit).is_err());
```

File: crates/rust-analyzer/src/handlers.rs [hunk 4]
```diff
@@ -463,8 +463,12 @@ pub(crate) fn handle_will_rename_files(
         .collect();
 
     // Drop file system edits since we're just renaming things on the same level
-    let edits = source_changes.into_iter().map(|it| it.source_file_edits).flatten().collect();
-    let source_change = SourceChange::from_edits(edits, Vec::new());
+    let mut source_changes = source_changes.into_iter();
+    let mut source_file_edits =
+        source_changes.next().map_or_else(Default::default, |it| it.source_file_edits);
```
Comment (matklad) [inline crates/rust-analyzer/src/handlers.rs:468]:
```suggestion
        source_changes.next().map(|it| it.source_file_edits).or_default();
```

File: crates/ide_db/src/source_change.rs (hunk 2 above)
Comment (Veykril) [reply crates/ide_db/src/source_change.rs]: Ye its actually not needed at all, thought `AssistBuilder` used it on its own but I didn't notice that it actually copied all of the `SourceChange` fields

**Review by matklad — COMMENTED** [review]
Negative diff, nice!

bors r+

**Pull request merged**